step-templates/mysql-backup-database.json

```diff
@@ -0,0 +1,105 @@
dulesPath -Force\n}\n\nfunction Get-NugetPackageProviderNotInstalled\n{\n\t# See if the nuget package provider has been installed\n    return ($null -eq (Get-PackageProvider -ListAvailable -Name Nuget -ErrorAction SilentlyContinue))\n}\n\nfunction Get-DatabaseExists\n{\n\t# Define parameters\n    param ($DatabaseName)\n    \n\t# Execute query\n    return Invoke-SqlQuery \"SHOW DATABASES LIKE '$DatabaseName';\"\n}\n\n# Define PowerShell Modules path\n$LocalModules = (New-Item \"$PSScriptRoot\\Modules\" -ItemType Directory -Force).FullName\n$env:PSModulePath = \"$LocalModules;$env:PSModulePath\"\n$PowerShellModuleName = \"SimplySql\"\n\n# Set secure protocols\n[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls11 -bor [System.Net.SecurityProtocolType]::Tls12\n\n# Check to see if SimplySql module is installed\nif ((Get-ModuleInstalled -PowerShellModuleName $PowerShellModuleName) -ne $true)\n{\n    # Tell user what we're doing\n    Write-Output \"PowerShell module $PowerShellModuleName is not installed, downloading temporary copy ...\"\n\n    # Install temporary copy\n    Install-PowerShellModule -PowerShellModuleName $PowerShellModuleName -LocalModulesPath $LocalModules\n}\n\n# Display\nWrite-Output \"Importing module $PowerShellModuleName ...\"\n\n# Check to see if it was downloaded\nif ((Test-Path -Path \"$LocalModules\\$PowerShellModuleName\") -eq $true)\n{\n\t# Import from temp location\n    $PowerShellModuleName =  \"$LocalModules\\$PowerShellModuleName\"\n}\n\n# Import the module\nImport-Module -Name $PowerShellModuleName\n\n# Create credential object for the connection\n$SecurePassword = ConvertTo-SecureString $createUserPassword -AsPlainText -Force\n$ServerCredential = New-Object System.Management.Automation.PSCredential ($createUsername, $SecurePassword)\n\ntry\n{\n\t# Connect to MySQL\n    $connectionString = \"Server=$createMySQLServerName;Port=$createPort;Uid=$createUsername;Pwd=$createUserPassword;\"\n    if ($createUseSSL -eq \"True\")\n    {\n\t\t# Append to connection string\n        $connectionString += \"SslMode=Required;\"\n    }\n    else\n    {\n    \t# Disable ssl\n        $connectionString += \"SslMode=none;\"\n    }\n    \n    Open-MySqlConnection -ConnectionString $connectionString\n\n    # See if database exists\n    $databaseExists = Get-DatabaseExists -DatabaseName $createDatabaseName\n\n    if ($databaseExists.ItemArray.Count -eq 0)\n    {\n        # Display message\n        Write-Error \"Database $createDatabaseName doesn't exist.\"\n \n    }\n    else\n    {\n\n        cd $createmysqlpath\n\n        $backupname = '{0}{1}-{2}.sql' -f ($CreateBackupDirectory, $createDatabaseName ,(Get-Date -Format FileDatetime))\n\n        .\\mysqldump.exe --databases $createDatabaseName > $backupname\n                \n        # Success\n        Write-Output \"$createDatabaseName was backed up!\"\n    }\n}\nfinally\n{\n    Close-SqlConnection\n}\n",
```

Comment:
**Questions:**
1. Function: `Close-SqlConnection` in the finally block. Is that defined by the powershell module `SimplySQL` ?
1. Should the `$backupname` variable also be a parameter so that the user can choose the name of the file?

Reply:
1. it's not explicitly stated to do that in Simply SQL
2. fair point however at some point in sharpening I'm going to extend this to give users more option so I'll look further then

Reply:
Cool, that sounds reasonable to me :+1: